WebXR User Manual: Anchors, Camera Color, DOM Overlay, Capabilities. (#578)
playcanvas/developer.playcanvas.com: docs/user-manual/xr/ar/anchors.md

---
title: Anchors
sidebar_position: 1
---

> Anchors provide the ability to create a point in 3D space that can be updated to match an ever-evolving understanding of real world by underlying the AR system. This allows for the placement of virtual objects in relation to the real world that feel planted in the user's environment.

Reviewer: ```suggestion
Anchors provide the ability to create a point in 3D space that can be updated to match an ever-evolving understanding of the real world by the underlying the AR system. This allows for the placement of virtual objects in relation to the real world that feel planted in the user's environment.
```

---
title: Anchors
sidebar_position: 1
---

> Anchors provides an ability to create a point in 3D space, that can be updated by to match ever evolving understanding of real world by underlying AR system. This allows to place virtual objects in relation to real world that feel planted in users environment.

Reviewer: ```suggestion
Anchors provide the ability to create a point in 3D space that can be updated to match an ever-evolving understanding of real world by underlying the AR system. This allows for the placement of virtual objects in relation to the real world that feel planted in the user's environment.
```

> Each anchor is represented as position and orientation, and can be created from arbitrary point as well as in relation to hit test result, that will make it more reliable.

Reviewer: ```suggestion
Each anchor is represented as a position and orientation and can be created from an arbitrary point as well as in relation to a hit test result that will make it more reliable.
```

        if (!err) {
            // new anchor has been created
        }
    });
});
```

## Anchor

Each anchor has its position and rotation and can be updated at any point. When an anchor is updated, the application developer should update related virtual objects accordingly.

Anchors can be added and removed dynamically during the session:

```javascript
app.xr.anchors.on('add', (anchor) => {
    const entity = new pc.Entity();

    // add a cone for an anchor
    entity.addComponent('render', { type: 'cone' });
    entity.setLocalScale(0.1, 0.1, 0.1); // 10cm diameter
    app.root.addChild(entity);

    // transform
    entity.setLocalPosition(anchor.getPosition());
    entity.setLocalRotation(anchor.getRotation());
    entity.translateLocal(0, 0.05, 0); // offset cone

    // update cone when anchor changes
    anchor.on('change', () => {
        entity.setLocalPosition(anchor.getPosition());
        entity.setLocalRotation(anchor.getRotation());
        entity.translateLocal(0, 0.05, 0); // offset cone
    });

    // remove cone when anchor is destroyed
    anchor.once('destroy', () => {
        entity.destroy();
    });
});
```

## Persistence

> Anchor persistence - provides a way to remember anchors between sessions, with a limited number of anchors per origin.
> This allows applications to place virtual objects in relation to the real-world geometry, and remain there between sessions.

Reviewer: ```suggestion
Anchor persistence provides a way to remember anchors between sessions, with a limited number of anchors per origin. This allows applications to place virtual objects in relation to the real-world geometry and remain there between sessions.
```

You can check if persistence is supported:

```javascript
if (app.xr.anchors.persistence) {
    // application can persist anchors
}
```

> Each anchor can have a UUID, that allows reference to it and restore between sessions.

Reviewer: ```suggestion
Each anchor can have a UUID that allows it to be referenced and restored between sessions.
```

> You can access a list of persistent anchors, and restore them on session start:

Reviewer: ```suggestion
You can access a list of persistent anchors and restore them on session start:
```
